src/hardware/gus.cpp

```diff
@@ -1,924 +1,1418 @@
+}
 
-	void generateSamples(Bit32s * stream,Bit32u len) {
-		//Disabled channel
-		if (RampCtrl & WaveCtrl & 3) return;
-
-		if (WaveCtrl & WCTRL_16BIT) {
-			for (int i = 0; i < (int)len; i++) {
-				// Get sample
-				Bit32s tmpsamp = GetSample16();
-				// Output stereo sample
-				stream[i << 1] += tmpsamp * VolLeft;
-				stream[(i << 1) + 1] += tmpsamp * VolRight;
-				WaveUpdate();
-				RampUpdate();
-			}
-		}
-		else {
-			for (int i = 0; i < (int)len; i++) {
-				// Get sample
-				Bit32s tmpsamp = GetSample8();
-				// Output stereo sample
-				stream[i << 1] += tmpsamp * VolLeft;
-				stream[(i << 1) + 1] += tmpsamp * VolRight;
-				WaveUpdate();
-				RampUpdate();
-			}
-		}
+void Gus::CheckVoiceIrq()
+{
+	irq_status &= 0x9f;
+	const Bitu totalmask = (voice_irq.vol_state | voice_irq.wave_state) &
+	                       active_voice_mask;
+	if (!totalmask)
+		return;
+	if (voice_irq.vol_state)
+		irq_status |= 0x40;
+	if (voice_irq.wave_state)
+		irq_status |= 0x20;
+	CheckIrq();
+	while (!(totalmask & 1ULL << voice_irq.status)) {
+		voice_irq.status++;
+		if (voice_irq.status >= active_voices)
+			voice_irq.status = 0;
 	}
-};
+}
 
-static GUSChannels *guschan[32];
-static GUSChannels *curchan;
-
-static void GUSReset(void) {
-	if((myGUS.gRegData & 0x1) == 0x1) {
-		// Reset
-		adlib_commandreg = 85;
-		myGUS.IRQStatus = 0;
-		myGUS.timers[0].raiseirq = false;
-		myGUS.timers[1].raiseirq = false;
-		myGUS.timers[0].reached = false;
-		myGUS.timers[1].reached = false;
-		myGUS.timers[0].running = false;
-		myGUS.timers[1].running = false;
-
-		myGUS.timers[0].value = 0xff;
-		myGUS.timers[1].value = 0xff;
-		myGUS.timers[0].delay = 0.080f;
-		myGUS.timers[1].delay = 0.320f;
-
-		myGUS.ChangeIRQDMA = false;
-		myGUS.mixControl = 0x0b;	// latches enabled by default LINEs disabled
-		// Stop all channels
-		int i;
-		for(i=0;i<32;i++) {
-			guschan[i]->RampVol=0;
-			guschan[i]->WriteWaveCtrl(0x1);
-			guschan[i]->WriteRampCtrl(0x1);
-			guschan[i]->WritePanPot(0x7);
-		}
-		myGUS.IRQChan = 0;
+uint32_t Gus::Dma8Addr()
+{
+	return static_cast<uint32_t>(dma_addr << 4);
+}
+
+uint32_t Gus::Dma16Addr()
+{
+	const auto lower = dma_addr & 0x1fff;
+	const auto upper = dma_addr & 0xc000;
+	const auto combined = (lower << 1) | upper;
+	return static_cast<uint32_t>(combined << 4);
+}
+
+bool Gus::PerformDmaTransfer()
+{
+	if (dma_channel->masked || !(dma_ctrl & 0x01))
+		return false;
+
+#if LOG_GUS
+	LOG_MSG("GUS DMA event: max %u bytes. DMA: tc=%u mask=0 cnt=%u",
+	        BYTES_PER_DMA_XFER, dma_channel->tcount ? 1 : 0,
+	        dma_channel->currcnt + 1);
+#endif
+
+	const auto addr = IsDmaXfer16Bit() ? Dma16Addr() : Dma8Addr();
+	const uint16_t desired = dma_channel->currcnt + 1;
+
+	if ((dma_ctrl & 0x2)) // Copy samples via DMA from GUS memory
+		dma_channel->Write(desired, ram + addr);
+
+	else if (!(dma_ctrl & 0x80)) // Skip DMA content
+		dma_channel->Read(desired, ram + addr);
+
+	else { // Copy samples via DMA into GUS memory
```

Comment:
Ending `else` has braces, therefore to be in accordance to K&R, all ifs should have braces.

Reply:
Got it; thanks!